ethereum/core/src/main/java/org/hyperledger/besu/ethereum/core/encoding/TransactionDecoder.java

```diff
@@ -50,154 +36,142 @@ interface Decoder {
+    return decodeTypedTransaction(typedTransactionBytes, transactionType, context);
   }
 
-  private static Decoder getDecoder(final TransactionType transactionType) {
-    return checkNotNull(
-        TYPED_TRANSACTION_DECODERS.get(transactionType),
-        "Developer Error. A supported transaction type %s has no associated decoding logic",
-        transactionType);
+  /**
+   * Decodes a typed transaction. The method first slices the transaction bytes to exclude the
+   * transaction type, then uses the appropriate decoder for the transaction type to decode the
+   * remaining bytes.
+   *
+   * @param transactionBytes the transaction bytes
+   * @param transactionType the type of the transaction
+   * @param context the encoding context
+   * @return the decoded transaction
+   */
+  private static Transaction decodeTypedTransaction(
+      final Bytes transactionBytes,
+      final TransactionType transactionType,
+      final EncodingContext context) {
+    // Slice the transaction bytes to exclude the transaction type and prepare for decoding
+    final RLPInput transactionInput = RLP.input(transactionBytes.slice(1));
+    // Use the appropriate decoder for the transaction type to decode the remaining bytes
+    return getDecoder(transactionType, context).decode(transactionInput);
   }
 
-  static Transaction decodeFrontier(final RLPInput input) {
-    input.enterList();
-    final Transaction.Builder builder =
-        Transaction.builder()
-            .type(TransactionType.FRONTIER)
-            .nonce(input.readLongScalar())
-            .gasPrice(Wei.of(input.readUInt256Scalar()))
-            .gasLimit(input.readLongScalar())
-            .to(input.readBytes(v -> v.size() == 0 ? null : Address.wrap(v)))
-            .value(Wei.of(input.readUInt256Scalar()))
-            .payload(input.readBytes());
-
-    final BigInteger v = input.readBigIntegerScalar();
-    final byte recId;
-    Optional<BigInteger> chainId = Optional.empty();
-    if (v.equals(REPLAY_UNPROTECTED_V_BASE) || v.equals(REPLAY_UNPROTECTED_V_BASE_PLUS_1)) {
-      recId = v.subtract(REPLAY_UNPROTECTED_V_BASE).byteValueExact();
-    } else if (v.compareTo(REPLAY_PROTECTED_V_MIN) > 0) {
-      chainId = Optional.of(v.subtract(REPLAY_PROTECTED_V_BASE).divide(TWO));
-      recId = v.subtract(TWO.multiply(chainId.get()).add(REPLAY_PROTECTED_V_BASE)).byteValueExact();
+  /**
+   * Decodes a transaction from opaque bytes. The method first determines the transaction type from
+   * the bytes. If the type is present, it delegates the decoding process to the appropriate decoder
+   * for that type. If the type is not present, it decodes the bytes as an RLP input.
+   *
+   * @param opaqueBytes the opaque bytes
+   * @param context the encoding context
+   * @return the decoded transaction
+   */
+  public static Transaction decodeOpaqueBytes(
+      final Bytes opaqueBytes, final EncodingContext context) {
+    var transactionType = getTransactionType(opaqueBytes);
```

Comment:
is this now doing the same as the decodeRLP, just a RLPInput instead of the bytes?